### PR TITLE
Add more rich network functions

### PR DESCRIPTION
**Pull request by WeiZhang555**
Implement more rich network functions
    
    * Support add ip addr to a specific network interface
    * Support delete ip address from a network interface
    * Support modify MTU
    * Support add any routes into container

### CODE REVIEW

File: hypervisor/network.go [hunk 1]
```diff
@@ -282,13 +284,61 @@ func (nc *NetworkContext) getInterface(id string) *InterfaceCreated {
 	return nil
 }
 
+func (nc *NetworkContext) AddIPAddr(id, ip string) error {
+	nc.slotLock.RLock()
```
Comment (gao-feng) [inline hypervisor/network.go]: Should not be read lock here.

File: hypervisor/network.go [hunk 2]
```diff
@@ -282,13 +284,61 @@ func (nc *NetworkContext) getInterface(id string) *InterfaceCreated {
...
+		return ErrNoSuchInf
+	}
+	inf.IpAddr = append(inf.IpAddr, ip)
+	inf.Desc.Ip = append(inf.Desc.Ip, ip)
+	return nil
+}
+
+func (nc *NetworkContext) DeleteIPAddr(id, ip string) error {
+	nc.slotLock.RLock()
+	defer nc.slotLock.RUnlock()
```
Comment (gao-feng) [inline hypervisor/network.go]: same here

File: hypervisor/network.go [hunk 3]
```diff
@@ -282,13 +284,61 @@ func (nc *NetworkContext) getInterface(id string) *InterfaceCreated {
...
+	inf, ok := nc.idMap[id]
+	if !ok {
+		return ErrNoSuchInf
+	}
+
+	for i, v := range inf.IpAddr {
+		if v == ip {
+			inf.IpAddr = append(inf.IpAddr[:i], inf.IpAddr[i+1:]...)
```
Comment (gao-feng) [inline hypervisor/network.go]: what happen when the last element in ipaddr array match the ip, is the inf.IpAddr[i+1] will access the array out of range?

File: hypervisor/network.go [hunk 4]
```diff
@@ -282,13 +284,61 @@ func (nc *NetworkContext) getInterface(id string) *InterfaceCreated {
...
+			break
+		}
+	}
+
+	for i, v := range inf.Desc.Ip {
+		if v == ip {
+			inf.IpAddr = append(inf.Desc.Ip[:i], inf.Desc.Ip[i+1:]...)
```
Comment (gao-feng) [inline hypervisor/network.go]: same here

File: hypervisor/network.go [hunk 5]
```diff
@@ -282,13 +284,61 @@ func (nc *NetworkContext) getInterface(id string) *InterfaceCreated {
...
+			break
+		}
+	}
+	return nil
```
Comment (gao-feng) [inline hypervisor/network.go]: return error or printout warning info if there is no matching ip?

File: hypervisor/network.go [hunk 6]
```diff
@@ -282,13 +284,61 @@ func (nc *NetworkContext) getInterface(id string) *InterfaceCreated {
...
+}
+
+func (nc *NetworkContext) UpdateMtu(id string, mtu uint64) error {
+	nc.slotLock.RLock()
+	defer nc.slotLock.RUnlock()
```
Comment (gao-feng) [inline hypervisor/network.go]: Seems like write lock is more appropriate here

File: hypervisor/network/network.go [hunk 7]
```diff
@@ -9,19 +9,20 @@ import (
...
+	Bridge    string
+	Device    string
+	File      *os.File
+	Automatic bool
 }
 
 const (
 	DefaultBridgeIface = "hyper0"
 	DefaultBridgeIP    = "192.168.123.0/24"
+	DefaultMtu         = 1450
```
Comment (gao-feng) [inline hypervisor/network/network.go]: why 1450 by default? it should not be 1500?

File: hypervisor/network/network_linux.go [hunk 8]
```diff
@@ -1049,45 +1049,27 @@ func AllocateAddr(requestedIP string) (*Settings, error) {
...
-		IPAddress:   ip.String(),
-		Gateway:     BridgeIPv4Net.IP.String(),
-		Bridge:      BridgeIface,
-		IPPrefixLen: maskSize,
-		Device:      "",
-		File:        nil,
-		Automatic:   true,
+		Mac:       mac,
+		IPAddress: []string{ip.String()},
+		Mtu:       DefaultMtu,
```
Comment (gao-feng) [inline hypervisor/network/network_linux.go]: I think by default we should not care about mtu, vm/os will setup the appropriate mtu for us. lets just set it null.

File: hypervisor/vm_states.go [hunk 9]
```diff
@@ -106,11 +108,85 @@ func (ctx *VmContext) restoreContainer(id string) (alive bool, err error) {
...
+		}
+		return nil
+	}
+}
+
+func (ctx *VmContext) hyperstartAddRoute(r []hyperstartapi.Route) error {
+	if err := ctx.hyperstart.AddRoute(r); err != nil {
+		return err
+	}
+	return nil
```
Comment (gao-feng) [inline hypervisor/vm_states.go]: return ctx.hyperstart.AddRoute(r)?

File: supervisor/proxy/nslistener.go [hunk 10]
```diff
@@ -213,19 +212,38 @@ func handleAddr(update netlink.AddrUpdate, callback func(supervisor.NetlinkUpdat
...
 
 	netlinkUpdate := supervisor.NetlinkUpdate{}
 	netlinkUpdate.Route = update
 	netlinkUpdate.UpdateType = supervisor.UpdateTypeRoute
 	callback(netlinkUpdate)
+
+	// delete all the routes from outside
+	// The reason that we need to delete all the routes is
+	// that if we don't do this, we will get two active network interfaces,
+	// and ping the IP will get duplicate ICMP replies.
```
Comment (gao-feng) [inline supervisor/proxy/nslistener.go]: The nic in netns of docker container(netns of nslistener) should be down, is it still has this problem?

File: hypervisor/network.go (hunk 1 above)
Comment (WeiZhang555) [inline hypervisor/network.go]: You're right.

File: hypervisor/network.go (hunk 3 above)
Comment (WeiZhang555) [inline hypervisor/network.go]: It should be fine, golang will handle it. See https://play.golang.org/p/5SWxZB3p2r

File: hypervisor/network/network.go (hunk 7 above)
Comment (WeiZhang555) [inline hypervisor/network/network.go]: I remember that vxlan interface has a mtu of 1450, once I met a problem that container network can't visit outside via vxlan, after setting its mtu to 1450, everything is good again.

But I'll keep mtu empty to let os handle it as your other comments suggest.

File: hypervisor/vm_states.go (hunk 9 above)
Comment (WeiZhang555) [inline hypervisor/vm_states.go]: Right :+1:

File: supervisor/proxy/nslistener.go (hunk 10 above)
Comment (WeiZhang555) [inline supervisor/proxy/nslistener.go]: I think we can still enter the netns and make the nic up? This should be a quite easy step, so we can't guarantee CNI plugin won't set nic up later.

Delete all route should be another protection for duplicate network flows.

Comment (gao-feng) [inline supervisor/proxy/nslistener.go]: I'm wonder if this deleting of router of netns will trigger route-update netlink event, though we don't care about this event right now.

Maybe removing the veth pair on host which joins the docker bridge is a good way, but the code will be complicated...

Comment (WeiZhang555) [inline supervisor/proxy/nslistener.go]: > Maybe removing the veth pair on host which joins the docker bridge is a good way

This is also another option, maybe we should do this?  For example, after tap created and inserted, removed veth from bridge?

**Comment by gnawux** [comment]
@WeiZhang555 could you help do a small rebase? The #537 introduced some conflicts

**Comment by WeiZhang555** [comment]
@gnawux I'll do it soon :wink:

File: supervisor/proxy/nslistener.go (hunk 10 above)
Comment (WeiZhang555) [inline supervisor/proxy/nslistener.go]: @gao-feng I just met a extremely weird problem, starting two container in one VM(POD), first one has proper route rules, but second container has no route rules. How could this happen? I thought they were in same network namespace and shares all routes.  Do you have any idea?

Comment (gao-feng) [inline supervisor/proxy/nslistener.go]: @WeiZhang555 
Can you check the ipaddress of nic? is it same on two container? they should stay in the same netns.

Comment (WeiZhang555) [inline supervisor/proxy/nslistener.go]: Yes, the nic ip is the same, only routes are different, quite weird. I'll try to do more investigation.

**Comment by WeiZhang555** [comment]
Closing this. We have a new one #597 carrying this :smile:

**Pull request closed without merging**